tests/apps/tests.py

```diff
@@ -528,3 +530,63 @@ def test_multiple_paths_explicit_path(self):
             with self.settings(INSTALLED_APPS=["nsapp.apps.NSAppConfig"]):
                 app_config = apps.get_app_config("nsapp")
                 self.assertEqual(app_config.path, self.app_path)
+
+
+class QueryPerformingAppTests(TestCase):
+    """
+    Tests related to accessing the database during apps.populate()
+    Refs #33143
+    """
+
+    databases = {"default", "other"}
+
+    expected_message = (
+        "Accessing the database during app initialization is discouraged."
+    )
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryDefaultDatabaseAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_when_importing_app_modules(self):
+        """
+        Trigger a warning when accessing the default database during apps.populate()
+        """
+        with self.clear_apps_registry():
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryOtherDatabaseAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_to_non_default_databases(self):
+        """
+        Trigger a warning when accessing a non-default database during apps.populate()
+        """
+        with self.clear_apps_registry():
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryUsingCursorAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_using_raw_cursors(self):
+        """
+        Trigger a warning when accessing the database using cursors during
+        apps.populate()
+        """
+        with self.clear_apps_registry():
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
+
+    @contextlib.contextmanager
+    def clear_apps_registry(self):
```

Comment:
Is there any specific reason to define this context manager instead of using `mock.patch`?